utils_linux.go

```
@@ -229,6 +229,14 @@ func (r *runner) run(config *specs.Process) (_ int, retErr error) {
 	if err != nil {
 		return -1, err
 	}
+	defer func() {
+		// We should terminate the process once we got an error.
+		// It's safe to do this because we ignored all errors in
+		// this function.
+		if retErr != nil {
+			r.terminate(process)
+		}
+	}()
```

Not sure I get this. Now this defer call will be used in places where we didn't have the `r.terminate()` call before. Like https://github.com/opencontainers/runc/pull/4661/commits/84ffd9953eff06ace2e21253d783c5253420cb8a#diff-bcda3d076af1f465c76f38c2404e7cd82b431bfd484d5c810871f9953162136bR254 , https://github.com/opencontainers/runc/pull/4661/commits/84ffd9953eff06ace2e21253d783c5253420cb8a#diff-bcda3d076af1f465c76f38c2404e7cd82b431bfd484d5c810871f9953162136bR265 and https://github.com/opencontainers/runc/pull/4661/commits/84ffd9953eff06ace2e21253d783c5253420cb8a#diff-bcda3d076af1f465c76f38c2404e7cd82b431bfd484d5c810871f9953162136bR272

Is this on purpose? Not sure what the "we ignored all errors in this function" means, like the callers of this function ignore the errors? Here we are not ignoring errors, we are returning errors in several cases.

What am I missing?

---

> Not sure what the "we ignored all errors in this function" means, like the callers of this function ignore the errors? Here we are not ignoring errors, we are returning errors in several cases.

We have ignored any errors here:
https://github.com/opencontainers/runc/blob/7f34471b1fe5a81a0d46e799cff93cac254dac23/utils_linux.go#L333-L336

If `process.ops == nil`, it will also return an error:
https://github.com/opencontainers/runc/blob/main/libcontainer/process.go#L148-L151


> Not sure I get this. Now this defer call will be used in places where we didn't have the `r.terminate()` call before.

As we have ignored any errors in `r.terminate(process)`, so we can call it safely in any places when got an error.

---

I don't understand why we wouldn't just add this defer after the process has been created. Can you please elaborate?

---

> I don't understand why we wouldn't just add this defer after the process has been created.

I think I don't understand your question here. I think I just has done what you want to do, I added this defer after the process has been created. Perhaps I misunderstood what you said. If yes, please tell me.